services/cancellation-request.js

```diff
@@ -0,0 +1,223 @@
+	 * Create a cancellation form (modal) and show it to the user.
+	 * !! The class ID section should not be edited by the user
+	 * 
+	 * @param {Interaction} interaction The user interaction object
+	 */
+	async showCancellationForm(interaction) 
+	{
+		// Retrieve the chosen class Id from the interaction
+        const classId = interaction.values[0];
+
+		// Create the modal object
+		const modal = new ModalBuilder()
+			.setCustomId('cancellationForm')
+			.setTitle('Class Cancellation Form');
+
+		// Add components to modal
+
+		// Create the cancellation reason input component
+		const feedback = new TextInputBuilder()
+			.setCustomId('cancellationReason')
+			.setLabel("Why would you like to cancel this class?")
+			.setPlaceholder('Because I have final exams coming up!')
+		    // Paragraph means multiple lines of text.
+			.setStyle(TextInputStyle.Paragraph);
+
+		// This text input holds the class ID, which the user shouldn't change
+		const classInfo = new TextInputBuilder()
+			.setCustomId('cancellationClassId')
+			.setLabel(`Leave as ${classId} unless instructed otherwise`)
+			.setPlaceholder(`Set as ${classId}`)
+			// This is the default value that will be inputted
+			.setValue(classId)
+			.setStyle(TextInputStyle.Short);
+
+		// An action row only holds one text input,
+		// so one action row per text input is needed.
+		const firstActionRow = new ActionRowBuilder().addComponents(feedback);
+		const secondActionRow = new ActionRowBuilder().addComponents(classInfo);
+
+		// Add inputs to the modal
+		modal.addComponents(firstActionRow, secondActionRow);
+
+		try {
+            // Show the modal to the user
+			await interaction.showModal(modal);
+        } catch (error) {
+            console.error(error);
+            await interaction.reply({ content: 'There was an error while executing this command!', ephemeral: true });
+        }
+	},
+
+    /**
+     * Handle the cancellation form submission.
+     * Extract the data (reason & classId) from the interaction, then make an
+     * HTTP POST request via the API to send the request to the server.
+     * 
+     * @param {Interaction} interaction The user interaction object
+     */
+    async cancellationFormSubmission(interaction)	{
+        // Get the data entered by the user
+        const cancellationReason = interaction.fields.getTextInputValue('cancellationReason');
+        const selectedClassId = interaction.fields.getTextInputValue('cancellationClassId');
+        
+        // Holds the extracted data in JSON format (to be sent to the API)
+        const requestData = {
+            class_ID: selectedClassId,
+            reason: cancellationReason
+        };
+
+		// The parameters of the HTTP POST request
+		const params = {
+			method: "POST",
+			body: JSON.stringify(requestData),
+        	headers : {'Content-Type': 'application/json'}
+        };
+
+		// The API URL for submitting class feedback
+		const url = "http://localhost:8080/cancellation_request";
```

Comment:
Just a not here: would be cool if all those url constants for the various calls are structure a bit differently by:
- having a const urlPrefix variable in the index.js file that will store http://localhost:8080 for now but that can be changed easily when we host it somewhere else
- having const url be equal to the urlPrefix + the endpoint so like: const url = urlPrefix + "/cancellation_request" in that case
-

Reply:
Hello Anton, that's a great idea.

I've added it to the config.json file instead of index.js so add the following to your config.json file:

"apiUrlPrefix" : "http://localhost:8080",